File: clipsync-win/ClipSync.Core/Clipboard/ForegroundRing.cs
using System;
using System.Collections.Generic;
using ClipSync.Settings;

namespace ClipSync.Clipboard;

/// A short, bounded history of which app was in the foreground and when.
///
/// Clipboard change notifications arrive asynchronously, so by the time
/// the watcher has an item the user may already have switched apps.
/// Recording transitions as they happen lets the watcher ask what was in
/// front at the moment of the copy rather than at the moment of handling.
public sealed class ForegroundRing
{
    public const int MaxEntries = 16;
    public static readonly TimeSpan MaxAge = TimeSpan.FromMinutes(2);

    private readonly List<Entry> _entries = new();
    private readonly object _lock = new();

    private readonly record struct Entry(DateTime At, AppIdentity? App);

    public void Record(DateTime atUtc, AppIdentity? app)
    {
        lock (_lock)
        {
            _entries.Add(new Entry(atUtc, app));
            Trim(atUtc);
        }
    }

    /// The app whose interval contains `utc`, or null if `utc` predates
    /// everything retained (which the caller treats as fail-open).
    public AppIdentity? AppAt(DateTime utc)
    {
        lock (_lock)
        {
            for (var i = _entries.Count - 1; i >= 0; i--)
                if (_entries[i].At <= utc) return _entries[i].App;
            return null;
        }
    }

    /// Caller holds _lock. Always leaves at least the newest entry, so a
    /// user who has sat in one app for hours still resolves.
    private void Trim(DateTime nowUtc)
    {
        // Find the newest expired entry, then remove everything up to and
        // including it in one shot — removing inside the scan would
        // invalidate the indices we are still iterating over.
        var lastExpired = -1;
        for (var i = 0; i < _entries.Count - 1; i++)
            if (nowUtc - _entries[i].At > MaxAge) lastExpired = i;
        if (lastExpired >= 0) _entries.RemoveRange(0, lastExpired + 1);

        while (_entries.Count > MaxEntries) _entries.RemoveAt(0);
    }
}

File: clipsync-mac/Sources/ClipSync/Clipboard/ForegroundRing.swift
import Foundation

/// A short, bounded history of which app was frontmost and when.
///
/// `PasteboardWatcher` polls `changeCount` every 200 ms, so the copy time
/// is only known to within that interval. The suppression rule therefore
/// needs "which apps held focus between t1 and t2", not just "which app
/// held focus at t" — a different query from the Windows ring, and the
/// reason `appsIn(_:)` exists.
///
/// Semantics mirror the Windows `ForegroundRing`: each entry owns the
/// half-open interval `[its timestamp, the next entry's timestamp)`, the
/// newest entry runs to infinity, and a timestamp falling exactly on a
/// transition resolves to the newly-activated app. Bounded in both size
/// and age; anything evicted reads back as unknown, which the decision
/// treats as fail-open (transmit).
final class ForegroundRing {
    static let maxEntries = 16
    static let maxAge: TimeInterval = 120

    private struct Entry {
        let at: Date
        let app: AppIdentity?    // nil = frontmost app could not be resolved
    }

    private var entries: [Entry] = []
    private let lock = NSLock()

    func record(at: Date, app: AppIdentity?) {
        lock.lock(); defer { lock.unlock() }
        entries.append(Entry(at: at, app: app))
        // Trim as a filter, not an in-place walk (the Windows first draft
        // mutated the list while iterating over it). Always keep the newest
        // entry, so a user who has sat in one app for hours still resolves.
        let cutoff = at.addingTimeInterval(-ForegroundRing.maxAge)
        if entries.count > 1 {
            let newest = entries.removeLast()
            entries = entries.filter { $0.at >= cutoff }
            entries.append(newest)
        }
        while entries.count > ForegroundRing.maxEntries {
            entries.removeFirst()
        }
    }

    /// The app whose interval contains `t`, or nil if `t` predates
    /// everything retained (fail open) or the entry recorded no identity.
    func appAt(_ t: Date) -> AppIdentity? {
        lock.lock(); defer { lock.unlock() }
        for entry in entries.reversed() where entry.at <= t {
            return entry.app
        }
        return nil
    }

    /// Every resolved app that held focus at any point in `(start, end]` —
    /// the polling window preceding a tick. An entry whose ownership
    /// interval `[at, next.at)` overlaps the window is included. Portions
    /// of the window older than the oldest retained entry are unknown and
    /// contribute nothing, which the caller treats as fail-open.
    func appsIn(start: Date, end: Date) -> [AppIdentity] {
        lock.lock(); defer { lock.unlock() }
        var found: [AppIdentity] = []
        for (i, entry) in entries.enumerated() {
            let intervalEnd = i + 1 < entries.count ? entries[i + 1].at : Date.distantFuture
            // Overlap of [entry.at, intervalEnd) with (start, end]:
            guard entry.at <= end, intervalEnd > start else { continue }
            if let app = entry.app, !found.contains(app) {
                found.append(app)
            }
        }
        return found
    }
}
